### PR TITLE
Add config option to disable dependency on the Gradle check task

### PR DESCRIPTION
**Pull request by atulgpt**
Fixes https://github.com/detekt/detekt/issues/5965

### CODE REVIEW

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt [hunk 1]
```diff
@@ -79,6 +79,8 @@ open class DetektExtension @Inject constructor(objects: ObjectFactory) : CodeQua
 
     var allRules: Boolean = DEFAULT_ALL_RULES_VALUE
 
+    var isEnforceCheck: Boolean = DEFAULT_IS_ENFORCE_CHECK
```
Comment (cortinico) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt:82]: This is not correct, it should be a `Property<Boolean>` with:

```
    val isEnforceCheck: Property<Boolean> =
      objects.property(Boolean::class.java).convention(DEFAULT_IS_ENFORCE_CHECK)
```

**Review by cortinico — CHANGES_REQUESTED** [review]

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt (hunk 1 above)
Comment (atulgpt) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/extensions/DetektExtension.kt]: Hi, @cortinico updated this part. Thanks for pointing out

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt [hunk 2]
```diff
@@ -142,6 +142,10 @@ abstract class Detekt @Inject constructor(
     @get:Internal
     internal abstract val ignoreFailuresProp: Property<Boolean>
 
+    @get:Input
+    @get:Optional
+    abstract val enforceCheck: Property<Boolean>
```
Comment (atulgpt) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt:147]: Hi, @cortinico / @3flex can we somehow rename this property to `enforceCheck`? I was trying to do so but `gradle` expects the Java bean convention and fails to compile the task. Using `isEnforceCheck` in Kotlin creates `isEnforceCheck()` getter which violates the bean naming convention

Also `@get:JvmName` doesn't work in abstract properties

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektPlain.kt [hunk 3]
```diff
@@ -26,7 +26,7 @@ internal class DetektPlain(private val project: Project) {
         }
 
         tasks.matching { it.name == LifecycleBasePlugin.CHECK_TASK_NAME }.configureEach {
-            if (extension.isEnforceCheck) {
+            if (extension.isEnforceCheck.getOrElse(DetektExtension.DEFAULT_IS_ENFORCE_CHECK)) {
```
Comment (cortinico) [inline detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektPlain.kt:29]: I belive here just `.get()` should be fine as teh default is set in the convention

Comment (atulgpt) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/internal/DetektPlain.kt]: Updated this

**Review by 3flex — CHANGES_REQUESTED** [review]
In the past I've considered implementing this not as an on/off option on the extension, but mirroring the way the Gradle code quality extensions allow this to be configured.

They use a [`sourceSets` option](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.CheckstyleExtension.html#org.gradle.api.plugins.quality.CheckstyleExtension:sourceSets) which allows defining:
> The source sets to be analyzed as part of the check and build tasks.

A toggle makes sense for Spotless because it's only adding a dependency for a single task, but our plugin has a task per sourceset for the "plain" detekt tasks, so configuring per sourceset would be much more flexible. That way you can configure to check main source set on check/build but not any of the test source sets for example, or however you'd like to set it up.

The type resolution tasks are different since they're linked to compilations, which often include multiple sourcesets, so we might need a different parameter for those.

**Comment by cortinico** [comment]
> They use a [`sourceSets` option](https://docs.gradle.org/current/dsl/org.gradle.api.plugins.quality.CheckstyleExtension.html#org.gradle.api.plugins.quality.CheckstyleExtension:sourceSets) which allows defining:

If I'm reading this right, it seems like they don't allow to run checkstyle without being connected to the `check`/`build` task (i.e. you specify the `sourceSets` to be inspected and they will run during `check`).

Our case is a bit different as we want to allow to run detekt even when not being hooked to `check` right?

File: detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt (hunk 2 above)
Comment (atulgpt) [reply detekt-gradle-plugin/src/main/kotlin/io/gitlab/arturbosch/detekt/Detekt.kt]: Hi, @3flex thanks for reviewing this PR. I have removed `enforceCheck` from `Detekt` task input.
I am still curious if renaming `enforceCheck` was possible or not. Let me know if you know some workaround

**Comment by atulgpt** [comment]
> but our plugin has a task per sourceset for the "plain" detekt tasks, so configuring per sourceset would be much more flexible.

Also, I think every single Gradle project only has one `check` task. Currently, only one `detekt` task is added to `check` task. Although I find one [issue](https://github.com/detekt/detekt/issues/3122) replacing `detekt` to `detektMain` but even after that only one task gets added to check

**Comment by 3flex** [comment]
> it seems like they don't allow to run checkstyle without being connected to the check/build task

It's just the task dependencies that are affected by that option:

https://github.com/gradle/gradle/blob/v8.1.1/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AbstractCodeQualityPlugin.java#L196-L224

The tasks are still created and enabled and can be run directly. They're just not added as dependents of `check`.

> I think every single Gradle project only has one check task. Currently, only one detekt task is added to check task.

Yes, that's how it is today, but I'm working on a proposed change to the plugin that would change this. Also - if you add the `sourceSets` option, then you could add additional detekt source set dependencies to the check task much more easily in the meantime.

**Comment by atulgpt** [comment]
Hi, @3flex I think when `sourceSet.getTaskName(taskBaseName, null);` returns `detekt` here https://github.com/gradle/gradle/blob/1cf537a851c635c364a4214885f8b9798051175b/subprojects/code-quality/src/main/groovy/org/gradle/api/plugins/quality/internal/AbstractCodeQualityPlugin.java#LL217C40-L217C82 then `check`  task will depend on `detekt` o/w not. Let me know if I am correct or not.

So you are suggesting if we want to remove `detekt` from `check` we need to set the `sourceSet` to empty for `detekt`. Right? I am still not clear on what the end-to-end config will look like.

> but I'm working on a proposed change to the plugin that would change this.

If in your work you are also addressing https://github.com/detekt/detekt/issues/5965 then feel free to assign this to yourself. And I will close this PR

**Comment by 3flex** [comment]
> Let me know if I am correct or not.

Detekt doesn't extend AbstractCodeQualityPlugin so the behaviour of that class isn't directly relevant to the current Detekt Gradle Plugin behaviour - apologies if I've misunderstood your comment.

The work I'm doing will eventually lead to a fix but it won't be ready that quickly since it's dependent on some changes being made for detekt 2.0. I think this PR is fine in the meantime though with the way the plugin is currently implemented.

**Comment by cortinico** [comment]
> it's just that the flag may get removed again quite quickly in favour of a new parameter.

I would vote to don't add it then as it's going to be yet another breaking change for the user which we'll have to explain how to migrate from

**Comment by BraisGabin** [comment]
I'm not 100% sure if that means closing this PR or not. But most likely it will not be in 1.23.0 so I changed the milestone.

**Comment by marschwar** [comment]
I don't think this PR will be necessary in 2.0 as we will remove the dependency on check altogether if I recall correctly.

**Comment by atulgpt** [comment]
Let me close this PR. Ideally, code from this PR should not be needed in the final solution. if needed @3flex can look into it

**Pull request closed without merging**